### PR TITLE
Fix reporting Unsubscribe from List event

### PR DESCRIPTION
**Pull request by ihor-sviziev**
Switch to Klaviyo Lists v2 API endpoint 
This endpoint triggers "Unsubscribed from %ListName%" event

### CODE REVIEW

File: Helper/Data.php [hunk 1]
```diff
@@ -178,24 +178,36 @@ public function subscribeEmailToKlaviyoList($email, $first_name=null, $last_name
...
+        ]);
 
-        $ch = curl_init();
-        curl_setopt($ch, CURLOPT_URL, 'https://a.klaviyo.com/api/v1/list/' . $list_id . '/members/exclude');
-        curl_setopt($ch, CURLOPT_POST, count($fields));
-        curl_setopt($ch, CURLOPT_POSTFIELDS, join('&', $fields));
-        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+        $curl = curl_init();
+        curl_setopt_array($curl, [
+            CURLOPT_URL => $url,
```
Comment (remstone7) [inline Helper/Data.php]: this is okay and nothing wrong as it's just using v2... but i would much prefer a full refactor to create an isolated method that would handle making api requests, you would pass some arguments like `path, method, body` and then we wouldn't need to use curl all over the place like it currently is

**Comment by ihor-sviziev** [comment]
@remstone7 I added fix.
I wanted to migrate to some magento class that doing what you need - but it doesn't support DELETE requsts O_o 
https://github.com/magento/magento2/pull/25742

File: Helper/Data.php [hunk 2]
```diff
@@ -228,4 +237,44 @@ protected function make_request($path, $params) {
...
+            CURLOPT_CUSTOMREQUEST => $method,
+            CURLOPT_POSTFIELDS => $encodedParams,
+            CURLOPT_USERAGENT => self::USER_AGENT,
+            CURLOPT_HTTPHEADER => [
+                'Content-Type: application/json',
+                'Content-Length: ' . strlen($encodedParams)
+            ],
+        ]);
+
+        // Submit the POST request
```
Comment (remstone7) [inline Helper/Data.php]: could we remove this comment or just make it:
```suggestion
        // Submit the request
```

Comment (ihor-sviziev) [reply Helper/Data.php]: Sure

Comment (ihor-sviziev) [reply Helper/Data.php]: Done

File: Helper/Data.php (hunk 1 above)
Comment (ihor-sviziev) [reply Helper/Data.php]: Fixed

File: Helper/Data.php [hunk 3]
```diff
@@ -142,60 +151,60 @@ public function getKlaviyoLists($api_key=null){
...
+     * @return bool|string
+     */
+    public function unsubscribeEmailFromKlaviyoList($email)
     {
-        $list_id = $this->getNewsletter();
-        $api_key = $this->getPrivateApiKey();
+        $listId = $this->getNewsletter();
+        $apiKey = $this->getPrivateApiKey();
 
+        $path = 'api/v2/list/' . $listId . '/subscribe';
```
Comment (remstone7) [inline Helper/Data.php]: we have a const for /subscribe, please use that

File: Helper/Data.php [hunk 4]
```diff
@@ -142,60 +151,60 @@ public function getKlaviyoLists($api_key=null){
...
-        if ($last_name) $properties['$last_name'] = $last_name;
+        if ($firstName) $properties['$first_name'] = $firstName;
+        if ($lastName) $properties['$last_name'] = $lastName;
         if ($source) $properties['$source'] = $source;
 
-        $properties_val = count($properties) ? json_encode(array('profiles' => $properties)) : '{}';
+        $propertiesVal = ['profiles' => $properties];
 
-        $url = "https://a.klaviyo.com/api/v2/list/" . $list_id . $opt_in . "?api_key=" . $api_key;
+        $path = "api/v2/list/" . $listId . $optInSetting . "?api_key=" . $apiKey;
```
Comment (remstone7) [inline Helper/Data.php]: one more just nit thing, could we make a const

`LIST_V2_API = 'api/v2/list/'`
then use `$path = self::LIST_V2_API . $listId.....`

File: Helper/Data.php (hunk 3 above)
Comment (ihor-sviziev) [reply Helper/Data.php]: OK

File: Helper/Data.php (hunk 4 above)
Comment (ihor-sviziev) [reply Helper/Data.php]: Fixed

File: Helper/Data.php (hunk 3 above)
Comment (ihor-sviziev) [reply Helper/Data.php]: Fixed

**Comment by ihor-sviziev** [comment]
In general I would like to extract all logic that sends requests to API to some model(s), but it definitely would be separate PR. Now my need - fix the issue

File: Helper/Data.php [hunk 5]
```diff
@@ -142,60 +152,60 @@ public function getKlaviyoLists($api_key=null){
...
-        $list_id = $this->getNewsletter();
-        $api_key = $this->getPrivateApiKey();
+        $listId = $this->getNewsletter();
+        $apiKey = $this->getPrivateApiKey();
 
+        $path = self::LIST_V2_API . $listId . self::API_SUBSCRIBE;
         $fields = [
-            'api_key=' . $api_key,
-            'email=' . urlencode($email),
+            'api_key' => (string)$apiKey,
```
Comment (remstone7) [inline Helper/Data.php]: sorry one more thing (but looks good 👍), i know our api accepts two different types of auth either utm param of api_key or api_key in the json, could we move this into the sendApiRequest and add it as a param?

Comment (ihor-sviziev) [reply Helper/Data.php]: Why we need it? I did it as showed in the docs example:
https://www.klaviyo.com/docs/api/v2/lists#delete-subscribe
image

Is there any issues or benefits of doing that?

Comment (ihor-sviziev) [reply Helper/Data.php]: Or you meant that we can add api key param inside `sendApiRequest`,
like 
```php
private function sendApiRequest(string $path, array $params, string $method = 'POST')
    {
        $params['api_key'] = $this->getPrivateApiKey(); // <!-- see to this line
        $url = self::KLAVIYO_HOST . $path;

        $curl = curl_init();
        ...

Comment (remstone7) [reply Helper/Data.php]: yes please to this one 👍

**Comment by remstone7** [comment]
just 1 conflict should be a quick fix -- also, we're just bringing this down to test today/tomorrow and will merge with out 1.1.7 release

File: Helper/Data.php (hunk 5 above)
Comment (ihor-sviziev) [reply Helper/Data.php]: Done

**Comment by ihor-sviziev** [comment]
@remstone7 I fixed conflict

**Comment by ihor-sviziev** [comment]
@remstone7 I see that you already released version 1.1.7 and 1.1.8, did you missed to merge this PR?

**Comment by remstone7** [comment]
Missed this but merging now, had an urgent fix we had to get out.  Itll go out in 1.1.9 which should be in the next couple weeks

**Pull request merged**